Cards/CardContainer.gd

```diff
@@ -10,93 +10,184 @@ class_name CardContainer
 ## TODO We need a global way to deal cards so we can deal cards from
 ## card effects (eg: draw 2 cards). There is also no concept of draw pile/discard pile.
 
+signal on_card_counts_updated
 
 @export var card_scene: PackedScene
 @export var total_hand_width: float = 100
 @export var card_hovered_offset: float = 100
 @export var card_queued_offset: float = 100
-@export var default_hand: Array[CardBase]
+@export var default_deck: Array[CardBase]
+@export var starting_hand_size: int = 5
+@export var max_hand_size: int = 10
 
-var cards: Array[CardWorld]
+var cards_in_hand: Array[CardWorld] = []
+var draw_pile: Array[CardBase] = []
+var discard_pile: Array[CardBase] = []
+var queued_card: CardWorld = null
 
 
 func _ready() -> void:
 	PhaseManager.on_phase_changed.connect(_on_phase_changed)
-	CardManager.successful_card_play.connect(_on_card_successful_play)
+	CardManager.set_card_container(self)
+	
+	draw_pile = default_deck.duplicate()
+	draw_pile.shuffle()
 
 
 func _process(_delta: float) -> void:
 	_update_card_positions()
 
 
-func deal_cards() -> void:
-	discard_all_cards()
-	for card_index: int in default_hand.size():
-		# create card and add to list
-		var card_instance: CardWorld = card_scene.instantiate()
-		add_child(card_instance)
-		cards.append(card_instance)
-		card_instance.init_card(default_hand[card_index])
-		
-		# bind mouse events
-		var card_click_handler: ClickHandler = card_instance.get_click_handler()
-		card_click_handler.on_click.connect(_on_card_clicked.bind(card_instance))
-		card_click_handler.on_mouse_hovering.connect(_on_card_hovering.bind(card_instance))
-		card_click_handler.on_unhover.connect(_on_card_unhovered.bind(card_instance))
+func set_queued_card(card: CardWorld) -> void:
+	queued_card = card
+
+
+func remove_queued_card() -> void:
+	discard_card(queued_card)
+	set_queued_card(null)
+
+
+func is_card_queued() -> bool:
+	return queued_card != null
+
+
+func deal_to_hand_limit() -> void:
+	# deal to our starting hand size, ignoring amount of cards we already have
+	var amount_to_draw: int = maxi(0, starting_hand_size - cards_in_hand.size())
+	draw_cards(amount_to_draw)
+
+
+func draw_cards(amount: int) -> void:
+	for card_index: int in amount:
+		# limit amount of cards to a maximum
+		if cards_in_hand.size() < max_hand_size:
+			_draw_card()
 
 
 func discard_all_cards() -> void:
-	for card: CardWorld in cards:
-		card.queue_free()
-	cards.clear()
+	while cards_in_hand.size() > 0:
+		_discard_last_card()
 
 
-func remove_card(card: CardWorld) -> void:
-	var card_index: int = cards.find(card)
-	cards[card_index].queue_free()
-	cards.remove_at(card_index)
+func discard_random_card(amount: int) -> void:
+	var valid_cards_to_discard: Array[CardWorld] = []
+	
+	for card in cards_in_hand:
```

Comment:
Yeah true, that's probably cleaner

Reply:
Just realized we don't even need to check the queued_card anymore since we're calling discard_random_card from `on_post_card_played()` in the CardBase. Just pushed a change that simplifies even further